File: Swift/002LongestSubstringWithoutRepeatingCharacters/Sources/LongestSubstringWithoutRepeatingCharacters/LongestSubstringWithoutRepeatingCharacters.swift
import Foundation

class LongestSubstringWithoutRepeatingCharacters {
  func lengthOfLongestSubstring(_ str: String) -> Int {
    guard !str.isEmpty else { return 0 }
    if str.count == 1 { return 1 }
    if Set(str).count == 1 { return 1 }
    var result = 0, left = 0, right = 0, setOfChars = Set<Character>()
    let strArray = Array(str)
    while left < strArray.count && right < strArray.count {
      if setOfChars.contains(strArray[right]) {
        setOfChars.remove(strArray[ left ])
        left += 1
      } else {
        setOfChars.insert(strArray[right])
        right += 1
        result = max(result, right - left)
      }
    }
    return result
  }
}

File: C++/002LongestSubstringWithoutRepeatingCharacters/LongestSubstringWithoutRepeatingCharacters/src/Longestsubstringwithoutrepeatingcharacters.cpp
#include "Longestsubstringwithoutrepeatingcharacters.h"
#include <iostream>

void Longestsubstringwithoutrepeatingcharacters::hello() {
  std::cout << "hello" << std::endl;
}

int Longestsubstringwithoutrepeatingcharacters::lengthOfLongestSubstring(string str) {
  if (str.empty())
    return 0;
  if (str.size() == 1)
    return 1;
  int result = 0;
  int left = 0;
  int right = 0;
  set<char> setOfChars;
  while (left < str.size() && right < str.size()) {
    if (setOfChars.find(str[right]) != setOfChars.end()) {
      setOfChars.erase(str[left]);
      left += 1;
    } else {
      setOfChars.insert(str[right]);
      right += 1;
      result = max(result, right - left);
    }
  }
  return result;
}
